include/ur_client_library/log.h

```diff
@@ -22,18 +22,85 @@
 #ifdef ROS_BUILD
 #include <console_bridge/console.h>
 
-#define LOG_DEBUG CONSOLE_BRIDGE_logDebug
-#define LOG_WARN CONSOLE_BRIDGE_logWarn
-#define LOG_INFO CONSOLE_BRIDGE_logInform
-#define LOG_ERROR CONSOLE_BRIDGE_logError
-#define LOG_FATAL CONSOLE_BRIDGE_logError
+#define URCL_LOG_DEBUG CONSOLE_BRIDGE_logDebug
+#define URCL_LOG_WARN CONSOLE_BRIDGE_logWarn
+#define URCL_LOG_INFO CONSOLE_BRIDGE_logInform
+#define URCL_LOG_ERROR CONSOLE_BRIDGE_logError
+#define URCL_LOG_FATAL CONSOLE_BRIDGE_logError
 
 #else
 
-#define LOG_DEBUG(format, ...) printf("[DEBUG]: " format "\n", ##__VA_ARGS__)
-#define LOG_WARN(format, ...) printf("[WARNING]: " format "\n", ##__VA_ARGS__)
-#define LOG_INFO(format, ...) printf("[INFO]: " format "\n", ##__VA_ARGS__)
-#define LOG_ERROR(format, ...) printf("[ERROR]: " format "\n", ##__VA_ARGS__)
-#define LOG_FATAL(format, ...) printf("[FATAL]: " format "\n", ##__VA_ARGS__)
+#define URCL_LOG_DEBUG(...) urcl::log(__FILE__, __LINE__, urcl::LogLevel::DEBUG, __VA_ARGS__)
+#define URCL_LOG_WARN(...) urcl::log(__FILE__, __LINE__, urcl::LogLevel::WARN, __VA_ARGS__)
+#define URCL_LOG_INFO(...) urcl::log(__FILE__, __LINE__, urcl::LogLevel::INFO, __VA_ARGS__)
+#define URCL_LOG_ERROR(...) urcl::log(__FILE__, __LINE__, urcl::LogLevel::ERROR, __VA_ARGS__)
+#define URCL_LOG_FATAL(...) urcl::log(__FILE__, __LINE__, urcl::LogLevel::FATAL, __VA_ARGS__)
 
 #endif
+
+namespace urcl
+{
+/*!
+ * \brief Different log levels
+ */
+enum class LogLevel
+{
+  DEBUG = 0,
+  INFO,
+  WARN,
+  ERROR,
+  FATAL,
+  NONE
+};
+
+/*!
+ * \brief Inherit from this class to change the behavior when logging messages.
+ */
+class LogHandler
+{
+public:
+  virtual ~LogHandler()
+  {
+  }
+  /*!
+   * \brief Function to log a message
+   *
+   * \param file The log message comes from this file
+   * \param line The log message comes from this line
+   * \param loglevel Indicates the severity of the log message
+   * \param log Log message
+   */
+  virtual void log(const char* file, int line, LogLevel loglevel, const char* log) = 0;
+};
+
+/*!
+ * \brief Register a new LogHandler object, for handling log messages.
+ *
+ * \param loghandler Pointer to the new object
+ */
+void registerLogHandler(LogHandler* loghandler);
```

Comment:
I've got it implemented locally, I can make a PR to your branch later today...

Reply:
That would be nice, thank you very much. 

I should have checked, so that maintainers should be able to commit directly to this branch, so I guess you don't have to make a PR.

Reply:
I created a PR nevertheless so you can check it.